### ISSUE TITLE
metrics for mixedinput classification

### ISSUE THREAD

**Issue by stas00** [fastai/fastai#658]
Someone raised the issue [here](http://forums.fast.ai/t/structured-learner-for-kaggle-titanic/8378/37) that we don't have metrics functions for mixedinput classification (and we don't have any notebooks demonstrating this class of DL). 

It appears that somewhat similar to `accuracy/accuracy_np` - we need multidimensional torch preds support (plus these come in as log_preds). I implemented a version that works, but need help with better naming, before submitting PR.

```
def recall_torch(preds, targs, thresh=0.5):
    pred_pos = torch.max(preds > thresh, dim=1)[1]
    tpos = torch.mul((targs.byte() == pred_pos.byte()), targs.byte())
    return tpos.sum()/targs.sum()

def precision_torch(preds, targs, thresh=0.5):
    pred_pos = torch.max(preds > thresh, dim=1)[1]
    tpos = torch.mul((targs.byte() == pred_pos.byte()), targs.byte())
    return tpos.sum()/pred_pos.sum()

def log_fbeta_torch(log_preds, targs, beta, thresh=0.5):
    assert beta > 0, 'beta needs to be greater than 0'
    beta2 = beta ** 2
    preds = torch.exp(log_preds)
    rec = recall_torch(preds, targs, thresh)
    prec = precision_torch(preds, targs, thresh)
    return (1 + beta2) * prec * rec / (beta2 * prec + rec)

def log_f1_torch(log_preds, targs, thresh=0.5): return log_fbeta_torch(log_preds, targs, 1, thresh)
```
usage:

```
m.fit(lr, 2, cycle_len=2, cycle_mult=3, metrics=[accuracy,log_f1_torch])
100% 8/8 [00:00<00:00, 12.92it/s]
epoch      trn_loss   val_loss   accuracy   log_f1_torch 
    0      0.472473   0.438977   0.780899   0.722966  
    1      0.429974   0.426494   0.814607   0.763082  
    2      0.417189   0.444517   0.820225   0.778917  
    3      0.372991   0.430631   0.825843   0.7864    
    4      0.323502   0.440252   0.831461   0.791569  
    5      0.284224   0.49835    0.837079   0.806387  
    6      0.251587   0.454803   0.814607   0.760272  
    7      0.23454    0.459216   0.814607   0.760272  

Out[44]:
[array([0.45922]), 0.8146067335364524, 0.760271861001706]
```

Thank you.

**Comment by stas00** [comment]
Also these would have the same problem with multidimensional deps torch vectors, and `accuracy_thresh` appears to miss argument or rely on some globals? odd:

```
def accuracy_thresh(thresh):
    return lambda preds,targs: accuracy_multi(preds, targs, thresh)

def accuracy_multi(preds, targs, thresh):
    return ((preds>thresh).float()==targs).float().mean()

def accuracy_multi_np(preds, targs, thresh):
    return ((preds>thresh)==targs).mean()
```

**Comment by jph00** [comment]
I guess we should follow the same template as `accuracy` et al - the version without a suffix is torch, and the np version has `_np`; this does require modifying the notebooks though, which is a bit of extra trouble!

**Comment by stas00** [comment]
Meanwhile, I checked the notebooks and submitted https://github.com/fastai/fastai/pull/661
I think more work will be needed to sort out the log- vs non-log- versions of `preds`.

**Comment by sgugger** [comment]
I think this can be closed for now since the PR was merged.

**Comment by stas00** [comment]
Thank you, @sgugger. in the future I can close such open issued myself, it just might not happen immediately, I usually do that once I see the PR merged. But feel free to beat me to it any time.
